### ISSUE TITLE
realm: event dependency optimization

### ISSUE THREAD

**Issue by streichler** [StanfordLegion/legion#823]
Right now, all Event preconditions are handled in a generic fashion, waiting for a CPU thread on the dependent operation's node to become aware that the precondition has been satisfied before starting the dependent operation.  In many cases, latency can be reduced by pushing the triggering operation further down the stack.  (For example, a CUDA task dependent on another CUDA task can be started with just an inter-stream fence to ensure ordering.)

The works breaks into four subcategories:
- [ ] Create a way for low-level completion information to be attached to an event and for dependent operations to query that information.
- [ ] Specifically go after the CUDA task -> CUDA task case mentioned above as the first optimized path.
- [ ] Try to think through any potential deadlock scenarios that come from starting the dependent task before its precondition is done executing.  (For example, it seems prudent to wait until the precondition has finished issuing work and is just waiting for deferred stuff to complete.)
- [ ] Figure out how we want to show this in profiles.  Are we ok with dependency edges going backwards in time?

**Comment by lightsighter** [comment]
For bullet 2: CUDA task to peer-to-peer memcpy (either by the copy engine or by a DMA kernel) is also important.

For bullet 3: I don't see any potential deadlocks at least in the CUDA case. You'll still have the CPU thread doing its work and making progress on triggering stuff. As long as you don't have that thread block waiting for any GPU operations you should be fine. I can certainly see cases where you might get a priority inversion that becomes a performance bug if you expose lower priority work to the CUDA driver before higher priority work triggered by the CPU thread, but that should only be a performance bug and not a deadlock. 

For bullet 4: is there a reason we can't push the profiling measurements down the stack too and do them with CUDA events so that we don't appear to backwards facing dependence edges? Feels like it would also have the additional benefit of making the profiling measurements more precise: we would actually see the gaps between the work from one an earlier GPU kernel finished and when work created by the CPU thread lagged behind dependent GPU kernels that started nearly immediately because we pushed them down to the CUDA driver.
